Services/Administration/OJS.Services.Administration.Business/Problems/ProblemsBusinessService.cs

```diff
@@ -235,6 +244,96 @@ public override async Task<ProblemAdministrationModel> Edit(ProblemAdministratio
         return model;
     }
 
+    public async Task<ProblemRetestValidationModel> ValidateRetest(int id)
+    {
+        var submissionsCount = await this.submissionsData.GetCountByProblem(id);
+
+        if (submissionsCount == 0)
+        {
+            return new ProblemRetestValidationModel
+            {
+                SubmissionsCount = 0,
+                AverageExecutionTime = 0,
+                RetestAllowed = false,
+                Message = "No submissions could be taken for validation."
+            };
+        }
+
+        // Take 3 submissions with different results so an average execution time can be obtained
+        var relevantSubmissions = this.submissionsData.GetAllByProblem(id)
+            .Where(s => s.IsCompiledSuccessfully && s.StartedExecutionOn.HasValue && s.CompletedExecutionOn.HasValue)
+            .GroupBy(s => s.Points)
+            .Select(g => g.FirstOrDefault())
+            .Take(3);
+
+        if (!relevantSubmissions.Any())
+        {
+            return new ProblemRetestValidationModel
+            {
+                SubmissionsCount = submissionsCount,
+                AverageExecutionTime = 0,
+                RetestAllowed = false,
+                Message = "No successfully compiled submissions could be taken for validation."
+            };
+        }
+
+        var averageTimeDifferenceInSeconds = relevantSubmissions
+            .AsEnumerable()
+            .Select(s => (s.CompletedExecutionOn.Value - s.StartedExecutionOn.Value).TotalSeconds)
+            .Average();
```

Comment:
Why not calculate this average without `.AsEnumerable` and on all compiled submissions for the problem. What is the purpose of the `relevantSubmissions.Any()` check? If the average is taken from the query above, if no submissions match the criteria, the value will be 0. Then we can replace the `.Any` with `average == 0`, if I'm not mistaken.

Reply:
The `relevantSubmissions.Any()` was implemented because a `System.InvalidOperationException: Sequence contains no elements` when doing average on empty collection of items, even if the queries are done like this:
```
var relevantSubmissions = this.submissionsData.GetAllByProblem(id)
            .Where(s => s.IsCompiledSuccessfully && s.StartedExecutionOn.HasValue && s.CompletedExecutionOn.HasValue)
            .GroupBy(s => s.Points)
            .Select(g => g.FirstOrDefault())
            .Take(3)
            .ToList();
            
var averageTimeDifferenceInSeconds = relevantSubmissions
            .Select(s => (s.CompletedExecutionOn.Value - s.StartedExecutionOn.Value).TotalSeconds)
            .Average();
```